azure-pipelines.yml

```diff
@@ -0,0 +1,159 @@
+variables:
+  dockerVersion: '19.03.5'
+  dotnetVersion: '3.1.101'
+
+trigger:
+- master
+- development
+
+pr:
+- master
+- development
+
+pool:
+  vmImage: windows-2019
+
+jobs:
+- job: build
+  displayName: Build database project
+
+  steps:
+  - task: VSBuild@1
+    displayName: 'Build database solution'
+    inputs:
+      solution: src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/NHSD.BuyingCatalogue.Ordering.OrderingDatabase.sqlproj
+
+  - task: CopyFiles@2
+    displayName: 'Copy Files to: $(build.artifactstagingdirectory)/database'
+    inputs:
+      SourceFolder: src
+      Contents: '**\*.dacpac'
+      TargetFolder: '$(build.artifactstagingdirectory)/database'
+      CleanTargetFolder: true
+
+  - task: PublishBuildArtifacts@1
+    displayName: 'Publish Artifact: database'
+    inputs:
+      PathtoPublish: '$(build.artifactstagingdirectory)/database'
+      ArtifactName: database
+
+  - task: colinsalmcorner.colinsalmcorner-buildtasks.dacpac-report-task.DacPacReport@1
+    displayName: 'DacPac Schema Compare NHSD.BuyingCatalogue.Ordering.OrderingDatabase.dacpac'
+    inputs:
+      dropName: database
+      dacpacName: NHSD.BuyingCatalogue.Ordering.OrderingDatabase
+      targetDacPacPath: '$(build.artifactstagingdirectory)/database'
+    enabled: false
+
+- job: dockerBuildAndPush
+  displayName: Run Unit & Integration tests and build & push docker containers to the acr
+
+  pool:
+    vmImage: 'ubuntu-latest'
+
+  steps:
+
+  - task: UseDotNet@2
+    displayName: 'Use DotNet Core $(dotnetVersion)'
+    inputs:
+      version: $(dotnetVersion)
+      includePreviewVersions: false
+
+  - task: DockerInstaller@0
+    inputs:
+      dockerVersion: $(dockerVersion)
+
+  - script: docker-compose up --build -d
+    displayName: 'Docker up'
+
+  - task: DotNetCoreCLI@2

```

Comment:
Just wanted to point your attention at the following resource [Docker with integration and unit tests](https://medium.com/@JoeHonour/a-guide-to-setting-up-a-net-core-project-using-docker-with-integrated-unit-and-component-tests-a326ca5a0284) as I think it maybe useful to run the unit tests from within the container. I have no evidence to base this off but I assume by running the run tests from the project would apply another restore & a build that should have already been carried out within the docker build. I think an improvement could be made here & potentially speed up the build process.

Reply:
Thank you for bringing this up. 

In essence, I think you're right with those assumptions and you're also right that it would speed up the build process.

However, this would hinder the development process, as running all tests when spinning up your environment would slow down the feedback loop, not to mention possible cases when you won't even be able to build due to failing tests.

So we have 2 things to weigh against each other:

| PROS                                      | CONS                       |
| :------------------------------: | :---------------------: |
| Faster completion of the pipeline process | Slower development process |

My opinion is that the overall time saved by faster build process will *not* outweigh the time lost on running tests locally when spinning up the environment.

I think it might be a good idea to get opinions from more people and see where the consensus lies and take action based on that.

What do you think? @sam-roberts-bjss